clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java

```diff
@@ -263,15 +677,56 @@ else if (value instanceof String)
         BOOLEAN, STRING, INT, SHORT, LONG, DOUBLE, LIST, CLASS, PASSWORD
     }
 
+    /**
+     * The importance level for a configuration
+     */
     public enum Importance {
         HIGH, MEDIUM, LOW
     }
 
     /**
-     * Validation logic the user may provide
+     * The width of a configuration value
+     */
+    public enum Width {
+        NONE, SHORT, MEDIUM, LONG
+    }
+
+    /**
+     * This is used by the {@link #validate(Map)} to get valid values for a configuration given the current
+     * configuration values in order to perform full configuration validation and visibility modification.
+     * In case that there are dependencies between configurations, the valid values and visibility
+     * for a configuration may change given the values of other configurations.
+     */
+    public interface Recommender {
+
+        /**
+         * The valid values for the configuration given the current configuration values.
+         * @param name The name of the configuration
+         * @param props The name of the configuration
+         * @return The list of valid values. To function properly, the returned objects should have the type
+         * defined for the configuration using the recommender.
+         */
+        List<Object> validValues(String name, Map<String, String> props);
```

Comment:
I don't think my question about this interface and it's use of the original `Map<String, String>` ever got answered. Doesn't using this mean we'll have to re-parse everything for every recommender? The only place this interface is used is after we've done all the parsing steps we would normally have done for a config. Couldn't we pass in the already parsed values given the current configuration? Is there a reason to use the raw properties instead of the parsed versions?